### PR TITLE
create Losses module

### PR DESCRIPTION
**Pull request by CarloLucibello**
Continuation of #1150, grouping losses under a `Losses` module as discussed. 

An alternative for the module name could be `Loss`, but `Losses` seemed more natural.

I also renamed `bce` back to `binarycrossentropy`, since now that the function is within a module I could provide I deprecation path without changing the function name with respect to last tagged release.

Some of the function contain the  `_loss` suffix, (e.g. `hinge_loss`, `poisson_loss`). We could drop that now that we have a namespace disambiguating the meaning, but again it seems more natural to keep, closer to the way people referer to them when speaking

### PR Checklist

- [x] Tests are added
- [x] Entry in NEWS.md
- [x] Documentation, if applicable
- [ ] Final review from `@MikeInnes` or `@dhairyagandhi96` (for API changes).

### CODE REVIEW

**Comment by johnnychen94** [comment]
> An alternative for the module name could be Loss, but Losses seemed more natural.

How about `Metrics`?

**Comment by CarloLucibello** [comment]
> How about Metrics as it can be a broader concept than Losses?

mhmh, don't really fancy that, I usually think of metrics as something you keep track of, not optimize

**Comment by DhairyaLGandhi** [comment]
`Metrics` sounds better to my ear as well, the metrics that you mentioned also just sounds like the tracking information that would just be used to see training progress,

**Comment by johnnychen94** [comment]
Disclaimer: I don't have a strong preference here. I proposed `Metrics` because

* it is not uncommon to track losses during training, using `Metrics` provides a place for other non-loss metrics, such as accuracy, psnr, ssim, etc.
* `Losses` has three `s` and I sometimes incorrectly type it as 😢 `loses`.
* `loss` is either a verb or a noun, while `metric` is mostly used as a noun. Using `metric`  as a module name has a clearer meaning.

**Comment by CarloLucibello** [comment]
Keras has two separate modules, Losses, and Metrics
https://keras.io/api/metrics/
"A metric is a function that is used to judge the performance of your model.
Metric functions are similar to loss functions, except that the results from evaluating a metric are not used when training the model. Note that you may use any loss functions as a metric function."

pytorch instead has only losses, with metrics typically hand-coded by users. 

We currently only have only loss function here, not arbitrary metrics, so I think `Losses` would be a more appropriate name, something which conveys a clear sense of the primary purpose. If the concern is about future expandability, we could add a `Metrics` module at any point, possibly re-exporting `Losses` as well.

**Comment by CarloLucibello** [comment]
@DhairyaLGandhi merge?

**Comment by CarloLucibello** [comment]
CI failure likely due to https://github.com/JuliaDiff/ChainRules.jl/issues/227

File: src/Flux.jl [hunk 1]
```diff
@@ -35,14 +35,16 @@ include("onehot.jl")
...
-include("layers/losses.jl")
 include("layers/basic.jl")
 include("layers/conv.jl")
 include("layers/recurrent.jl")
 include("layers/normalise.jl")
 
 include("data/Data.jl")
 
+include("losses/Losses.jl")
+using .Losses # TODO: stop importing Losses in Flux's namespace in v0.12
```
Comment (DhairyaLGandhi) [inline src/Flux.jl:46]: Loss should export all the functions as before.

File: src/losses/Losses.jl [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+mae(ŷ, y; agg=mean) = agg(abs.(ŷ .- y))
+
+"""
+    mse(ŷ, y; agg=mean)
+
+Return the loss corresponding to mean square error: 
+    
+    agg((ŷ .- y).^2)
+"""
+mse(ŷ, y; agg=mean) = agg((ŷ .- y).^2)
```
Comment (DhairyaLGandhi) [inline src/losses/Losses.jl:38]: Since there is a regression wrt to performance, maybe we should consider removing the aggregation for now

File: test/cuda/cuda.jl [hunk 3]
```diff
@@ -4,53 +4,37 @@ using Flux: cpu, gpu
...
 using LinearAlgebra: I, cholesky, Cholesky
 
-@info "Testing GPU Support"
-
 @testset "CUDA" begin
+  x = randn(5, 5)
+  cx = gpu(x)
+  @test cx isa CuArray
 
-CUDA.allowscalar(false)
```
Comment (DhairyaLGandhi) [inline test/cuda/cuda.jl:11]: Why is this PR moving around tests? The tests should be largely untouched and esp tests unrelated to losses should be untouched

**Comment by DhairyaLGandhi** [comment]
The package shouldn't move the definition of losses to a module like so. See how Optimisers.jl handles the directory structure.

**Comment by DhairyaLGandhi** [comment]
The package shouldn't move the definition of losses to a module like so. See how Optimisers.jl handles the directory structure. I would also consider calling this package Loss.jl or similar otherwise it sounds weird.

File: src/Flux.jl (hunk 1 above)
Comment (CarloLucibello) [reply src/Flux.jl:46]: It is, that is no loss is exported but all live in the Flux namespace, as they currently are. The comment is about keep the losses in Flux.Losses namespace, something we may want to do in v0.12

File: src/losses/Losses.jl (hunk 2 above)
Comment (CarloLucibello) [reply src/losses/Losses.jl:38]: if you refer to #1255, `mean` is not affected by it

File: test/cuda/cuda.jl (hunk 3 above)
Comment (CarloLucibello) [reply test/cuda/cuda.jl:11]: I had to reorganize a bit the cuda tests, in order to reflect the presence of the new `Losses` module and the new test/cuda/losses.jl file

File: src/losses/Losses.jl [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    tversky_loss,
+    dice_coeff_loss,
+    poisson_loss,
+    hinge_loss, squared_hinge_loss
+
+include("utils.jl")
+include("functions.jl")
+
+# group here all losses, used in tests
+const ALL_LOSSES = [mse, mae, msle,
```
Comment (DhairyaLGandhi) [inline src/losses/Losses.jl:25]: Should be moved over to the tests instead.

File: src/losses/Losses.jl (hunk 2 above)
Comment (DhairyaLGandhi) [reply src/losses/Losses.jl]: I mean specifically in the backwards pass.

File: src/losses/Losses.jl (hunk 4 above)
Comment (CarloLucibello) [reply src/losses/Losses.jl:25]: I wasn't sure where to put it in tests, since it's used in two separate files

Comment (DhairyaLGandhi) [reply src/losses/Losses.jl:25]: Before the first stateless layer test is run.

**Comment by CarloLucibello** [comment]
> I would also consider calling this package Loss.jl or similar otherwise it sounds weird to my ear

I fear using the singular here may go against julia's guidelines for module naming. Any opinion from @MikeInnes @oxinabox ?

File: src/losses/Losses.jl (hunk 2 above)
Comment (CarloLucibello) [reply src/losses/Losses.jl]: which regression?

**Comment by DhairyaLGandhi** [comment]
We should also make sure that we don't message things incorrectly. Currently it seems that we are saying that these bit of code are the loss functions which are somehow separate from being some generic transform that we can differentiate through. Continuing calling them stateless layers makes that a little bit more explicit. This is an important distinction to make for a newcomer to our ecosystem.

File: src/losses/Losses.jl (hunk 2 above)
Comment (DhairyaLGandhi) [reply src/losses/Losses.jl]: This is with Flux@0.10.4. `cmse` is the one from master, `mymse` is a current implementation that I am working on. 

image

Comment (CarloLucibello) [reply src/losses/Losses.jl]: ok. But we can't remove aggregation now and add it back later, that's a lot of breakage. We should just improve performances under the hood. If you have a better implementation for `agg=mean` we can add a branch for that. But really for such basic functions as `mean((x .- x).^2)` we should expect Zygote to be perfomant, and track down the issue if it is not

File: src/losses/Losses.jl (hunk 4 above)
Comment (CarloLucibello) [reply src/losses/Losses.jl]: done

**Comment by DhairyaLGandhi** [comment]
bors try

**Comment by DhairyaLGandhi** [comment]
Hopefully we haven't missed any tests
